### PR TITLE
fix(config): replace default constants with default factories

### PR DESCRIPTION
**Pull request by mdaniline**
## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by viceice — APPROVED** [review]

**Review by viceice — COMMENTED** [review]
needs tests to resolve failing code coverage

**Review by viceice — CHANGES_REQUESTED** [review]
needs test for code coverage

File: lib/config/defaults.spec.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+          description: "thing",
+          name: "thing"
+        }
+        const val = getDefault(option);
+
+        expect(val).toBe(null);
+      });
+    });
+  });
+});
```
Comment (viceice) [inline lib/config/defaults.spec.ts:45]: missing trailing newline, run prettier

**Review by viceice — COMMENTED** [review]
I think just handling the array issue instead of changing all is a much simpler and better solution.

@secustor @rarkins WDYT?

**Comment by mdaniline** [comment]
FWIW I wrote it this way to avoid having specialised logic per type - it's pretty obvious what happens without needing to worry about edge cases. Also it leaves this future-proofed in case we ever want to return `{}` instead of `null` for objects.

File: lib/config/defaults.spec.ts (hunk 1 above)
Comment (mdaniline) [reply lib/config/defaults.spec.ts]: Done

**Review by viceice — DISMISSED** [review]

File: lib/config/defaults.spec.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    ['string', 'object', 'integer'].forEach((type: unknown) => {
+      it(`returns null for ${type} values`, () => {
+        const option: RenovateOptions = {
+          type: type as 'string' | 'object' | 'integer',
+          description: 'thing',
+          name: 'thing',
+        };
+        const val = getDefault(option);
+
+        expect(val).toBe(null);
```
Comment (secustor) [inline lib/config/defaults.spec.ts:40]:
```suggestion
        expect(val).toBeNull();
```

File: lib/config/defaults.spec.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        type: 'boolean',
+        description: 'thing',
+        name: 'thing',
+      };
+      const val = getDefault(option);
+
+      expect(val).toBe(true);
+    });
+
+    ['string', 'object', 'integer'].forEach((type: unknown) => {
```
Comment (secustor) [inline lib/config/defaults.spec.ts:31]:
```suggestion
    ['string', 'object', 'integer'].forEach((type: string) => {
```

**Review by secustor — APPROVED** [review]

File: lib/config/defaults.spec.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        type: 'boolean',
+        description: 'thing',
+        name: 'thing',
+      };
+      const val = getDefault(option);
+
+      expect(val).toBe(true);
+    });
+
+    ['string', 'object', 'integer'].forEach((type: string) => {
```
Comment (viceice) [inline lib/config/defaults.spec.ts:31]: @zharinov can you convert to jest each function instead?

**Pull request merged**